server/app/services/applicant/ServiceAreaUpdateResolver.java

```diff
@@ -0,0 +1,187 @@
+            .collect(ImmutableList.toImmutableList());
+
+    if (filteredServiceAreaOptions.size() == 0) {
+      // return an update with the existing service areas
+      return CompletableFuture.completedFuture(
+          Optional.of(
+              ServiceAreaUpdate.create(serviceAreaPath, existingServiceAreaInclusionGroup)));
+    }
+
+    AddressLocation addressLocation = getAddressLocationFromUpdates(addressQuestion, updateMap);
+
+    ImmutableList<CompletionStage<ImmutableList<ServiceAreaInclusion>>>
+        serviceAreaInclusionGroupFutures =
+            filteredServiceAreaOptions.stream()
+                .map(
+                    (option) -> {
+                      return esriClient
+                          .getServiceAreaInclusionGroup(option, addressLocation)
+                          .thenApplyAsync(
+                              (serviceAreaInclusionGroup) -> serviceAreaInclusionGroup,
+                              httpExecutionContext.current());
+                    })
+                .collect(ImmutableList.toImmutableList());
+
+    return CompletableFuture.allOf(
+            serviceAreaInclusionGroupFutures.toArray(
+                new CompletableFuture[serviceAreaInclusionGroupFutures.size()]))
+        .thenComposeAsync(
+            (u) -> {
+              ImmutableList<ServiceAreaInclusion> serviceAreaInclusionGroup =
+                  serviceAreaInclusionGroupFutures.stream()
+                      .map((future) -> future.toCompletableFuture().join())
+                      .flatMap(Collection::stream)
+                      .collect(ImmutableList.toImmutableList());
+              ImmutableList<ServiceAreaInclusion> newServiceAreaInclusionGroup;
+
+              if (existingServiceAreaInclusionGroup.size() > 0) {
+                newServiceAreaInclusionGroup =
+                    Stream.of(existingServiceAreaInclusionGroup, serviceAreaInclusionGroup)
+                        .flatMap(List::stream)
+                        .collect(
+                            Collectors.toMap(
+                                ServiceAreaInclusion::getServiceAreaId,
+                                area -> area,
+                                (ServiceAreaInclusion existingInclusion,
+                                    ServiceAreaInclusion newInclusion) ->
+                                    newInclusion == null ? existingInclusion : newInclusion))
+                        .values()
+                        .stream()
+                        .collect(ImmutableList.toImmutableList());
+              } else {
+                newServiceAreaInclusionGroup = serviceAreaInclusionGroup;
+              }
+
+              ServiceAreaUpdate serviceAreaUpdate =
+                  ServiceAreaUpdate.create(serviceAreaPath, newServiceAreaInclusionGroup);
+
+              return CompletableFuture.completedFuture(Optional.of(serviceAreaUpdate));
```

Comment:
**suggestion (blocking):** There's nothing async happening inside this lambda. If you change L115 to `.thenAppyAsync` this line can be:

```suggestion
              return Optional.of(ServiceAreaUpdate.create(serviceAreaPath, newServiceAreaInclusionGroup));
```

Reply:
updated